packages/rpc-provider/src/substrate-connect/ScProvider.ts

```diff
@@ -0,0 +1,346 @@
+// Copyright 2017-2022 @polkadot/rpc-provider authors & contributors
+// SPDX-License-Identifier: Apache-2.0
+
+import type { JsonRpcResponse, ProviderInterface, ProviderInterfaceCallback, ProviderInterfaceEmitCb, ProviderInterfaceEmitted } from '../types';
+
+import { Chain, createScClient, ScClient, WellKnownChain } from '@substrate/connect';
+import EventEmitter from 'eventemitter3';
+
+import { assert, isError } from '@polkadot/util';
+
+import { RpcCoder } from '../coder';
+import { healthChecker } from './Health';
+
+type ResponseCallback = (response: string | Error) => void
+
+// These methods have been taken from:
+// https://github.com/paritytech/smoldot/blob/17425040ddda47d539556eeaf62b88c4240d1d42/src/json_rpc/methods.rs#L338-L462
+// It's important to take into account that smoldot is adding support to the new
+// json-rpc-interface https://paritytech.github.io/json-rpc-interface-spec/
+// However, at the moment this list only includes methods that belong to the "old" API
+const subscriptionUnsubscriptionMethods = new Map<string, string>([
+  ['author_submitAndWatchExtrinsic', 'author_unwatchExtrinsic'],
+  ['chain_subscribeAllHeads', 'chain_unsubscribeAllHeads'],
+  ['chain_subscribeFinalizedHeads', 'chain_unsubscribeFinalizedHeads'],
+  ['chain_subscribeFinalisedHeads', 'chain_subscribeFinalisedHeads'],
+  ['chain_subscribeNewHeads', 'chain_unsubscribeNewHeads'],
+  ['chain_subscribeNewHead', 'chain_unsubscribeNewHead'],
+  ['chain_subscribeRuntimeVersion', 'chain_unsubscribeRuntimeVersion'],
+  ['subscribe_newHead', 'unsubscribe_newHead'],
+  ['state_subscribeRuntimeVersion', 'state_unsubscribeRuntimeVersion'],
+  ['state_subscribeStorage', 'state_unsubscribeStorage']
+]);
+
+const wellKnownChains = new Set(Object.values(WellKnownChain));
+const scClients = new WeakMap<ScProvider, ScClient>();
+
+export class ScProvider implements ProviderInterface {
+  readonly #coder: RpcCoder = new RpcCoder();
+  readonly #spec: string | WellKnownChain;
+  readonly #dependency?: ScProvider;
+  readonly #subscriptions: Map<
+  string,
+  [ResponseCallback, { unsubscribeMethod: string; id: string | number }]
+  > = new Map();
+
+  readonly #requests: Map<number, ResponseCallback> = new Map();
+  readonly #eventemitter: EventEmitter = new EventEmitter();
+  #chain: Promise<Chain> | null = null;
+  #isChainReady = false;
+
+  public constructor (spec: string | WellKnownChain, dependency?: ScProvider) {
```

Comment:
This looks good but would rename `dependency` to `sharedSandbox`

Reply:
I think that I understand where this is coming from. However, from the point of view of the consumer, this is a dependency. As in the connection of this Provider that I'm creating depends on the connection of this other Provider... The fact that they have a shared sandbox behind the scenes it's a very important implementation detail, but I'm of the opinion that the end user doesn't have to know about it. Although, in the end this should be your call, so if you disagree with this reasoning I will rename it to `sharedSandbox`, for sure!

Reply:
Please rename to `sharedSandbox`, it's not just a parachain <-> relay chain thing. Sorry for the lack of deep explanation but I've explained this at least twenty times to many different people.

Reply:
done!